### PR TITLE
Scaling for a Deployment, ReplicaSet, Replication Controller, or Job.

### PR DESCRIPTION
**Pull request by kenan435**
Fixes https://github.com/kubernetes/dashboard/issues/1560

Scale or set a new size for a Deployment, ReplicaSet, Replication Controller. In the case of a job it will adjust the parallelism for the said job.

### CODE REVIEW

File: src/app/frontend/common/scaling/scaleresource_controller.js [hunk 1]
```diff
@@ -66,47 +60,77 @@ export default class UpdateReplicasDialogController {
...
+            transformRequest: function(headers) {
+              headers = angular.extend({}, headers, {'Content-Type': 'application/json'});
+              return angular.toJson(headers);
+            },
+          },
+        });
+  }
+
+  isJob() {
+    if (this.resourceKindName.toLowerCase() === 'job') {
```
Comment (maciaszczykm) [inline src/app/frontend/common/scaling/scaleresource_controller.js]: `return this.resourceKindName.toLowerCase() === 'job'` would do same job.

File: src/test/frontend/replicationcontroller/detail/updatereplicas_controller_test.js [hunk 2]
```diff
@@ -45,50 +47,52 @@ describe('Update Replicas controller', () => {
       state = $state;
       resource = $resource;
       httpBackend = $httpBackend;
-      httpBackend.expectGET('api/v1/csrftoken/replicationcontroller')
-          .respond(200, '{"token": "x"}');
+      // httpBackend.expectGET('api/v1/csrftoken/replicationcontroller')
```
Comment (maciaszczykm) [inline src/test/frontend/replicationcontroller/detail/updatereplicas_controller_test.js]: Remove it if it's not used.

File: src/app/backend/resource/scaling/scale.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+	rc = new(ReplicaCounts)
+	s, err := client.Extensions().Scales(namespace).Get(kind, name)
+	if err != nil {
+		return nil, err
+	}
+
+	rc.DesiredReplicas = s.Spec.Replicas
+	rc.ActualReplicas = s.Status.Replicas
+
+	return
```
Comment (maciaszczykm) [inline src/app/backend/resource/scaling/scale.go]: Can you test it to improve coverage?

File: src/app/backend/resource/scaling/scale.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+func ScaleResource(client client.Interface, kind, namespace, name, count string) (rc *ReplicaCounts, err error) {
+	rc = new(ReplicaCounts)
+
+	fmt.Println(kind)
```
Comment (kenan435) [inline src/app/backend/resource/scaling/scale.go]: Took this out. left it in there by accident.

**Comment by kenan435** [comment]
I know the build is failing 'cuz I am using normal binding:{{$ctrl...}} instead of one time binding: {{::$ctrl...}} How is this making the build fail? Any suggestions?

File: src/app/frontend/common/scaling/scaleresource_controller.js [hunk 5]
```diff
@@ -66,47 +60,73 @@ export default class UpdateReplicasDialogController {
...
+            method: 'PUT',
+            transformRequest: function(headers) {
+              headers = angular.extend({}, headers, {'Content-Type': 'application/json'});
+              return angular.toJson(headers);
+            },
+          },
+        });
+  }
+
+  isJob() {
```
Comment (maciaszczykm) [inline src/app/frontend/common/scaling/scaleresource_controller.js]: It has to be `@export`ed. You need annotation for this file too.

Comment (maciaszczykm) [inline src/app/frontend/common/scaling/scaleresource_controller.js]: Also try to shorten file names, ie. if there is only one module in directory, then you can name it `module.js`. Same with services or controllers.

**Comment by maciaszczykm** [comment]
> @maciaszczykm I know the build is failing 'cuz I am using normal binding:{{$ctrl...}} instead of one time binding: {{::$ctrl...}} How is this making the build fail since the code is valid? Any suggestions?

We use our own system for translations and it seems, that we have to use one-time bindings (didn't investigate it right now though). Most common pattern, that we are using when we want to live-update message is usage of i18n function from the controller.

**Comment by maciaszczykm** [comment]
> We use our own system for translations and it seems, that we have to use one-time bindings (didn't investigate it right now though). Most common pattern, that we are using when we want to live-update message is usage of i18n function from the controller.

Probably it happens due to fact, that if data is not one-time bound `google-closure-compiler` looks for value to be bound: https://github.com/bramstein/closure-compiler-inline/blob/master/src/com/google/javascript/jscomp/JsMessageVisitor.java#L534.

To provide such value you need to use similar call:

``` js
getEditTooltip() {
  /**
   * @type {string} @desc Generic "Edit some resource" tooltip text which appears over the
   * edit icon on the global action bar.
   */
  let MSG_ACTION_BAR_EDIT_TOOLTIP =
      goog.getMsg('Edit {$resourceName}', {'resourceName': this.resourceKindName});
  return MSG_ACTION_BAR_EDIT_TOOLTIP;
}
```

**Comment by maciaszczykm** [comment]
> @maciaszczykm I know the build is failing 'cuz I am using normal binding:{{$ctrl...}} instead of one time binding: {{::$ctrl...}} How is this making the build fail since the code is valid? Any suggestions?

We use our own system for translations and it seems, that we have to use one-time bindings. Probably it happens due to fact, that if data is not one-time bound `google-closure-compiler` looks for value to be bound: https://github.com/bramstein/closure-compiler-inline/blob/master/src/com/google/javascript/jscomp/JsMessageVisitor.java#L534.

Most common pattern, that we are using when we want to live-update message is usage of i18n function from the controller. To provide such value you need to use similar call:

``` js
getEditTooltip() {
  /**
   * @type {string} @desc Generic "Edit some resource" tooltip text which appears over the
   * edit icon on the global action bar.
   */
  let MSG_ACTION_BAR_EDIT_TOOLTIP =
      goog.getMsg('Edit {$resourceName}', {'resourceName': this.resourceKindName});
  return MSG_ACTION_BAR_EDIT_TOOLTIP;
}
```

File: src/app/frontend/replicaset/list/cardlist_component.js [hunk 6]
```diff
@@ -29,6 +29,11 @@ export class ReplicaSetCardListController {
 
     /** @private {!./../../common/namespace/service.NamespaceService} */
     this.kdNamespaceService_ = kdNamespaceService;
+
+    /** @private
+     * {!./../replicationcontrollerdetail/replicationcontroller_service.ReplicationControllerService}
```
Comment (floreks) [inline src/app/frontend/replicaset/list/cardlist_component.js]: Wrong annotation

File: src/app/frontend/replicaset/list/cardlist_component.js [hunk 7]
```diff
@@ -20,7 +20,7 @@ export class ReplicaSetCardListController {
    * @param {!./../../common/namespace/service.NamespaceService} kdNamespaceService
    * @ngInject
    */
-  constructor(kdNamespaceService) {
+  constructor(kdNamespaceService, kdScaleService) {
```
Comment (floreks) [inline src/app/frontend/replicaset/list/cardlist_component.js]: Missing annotation for scale service

File: src/app/frontend/replicaset/list/card_component.js [hunk 8]
```diff
@@ -42,6 +42,8 @@ export default class ReplicaSetCardController {
 
     /** @private {!./../../common/namespace/service.NamespaceService} */
     this.kdNamespaceService_ = kdNamespaceService;
+
+    this.kdScaleService_ = kdScaleService;
```
Comment (floreks) [inline src/app/frontend/replicaset/list/card_component.js]: missing annotation

File: src/app/frontend/replicaset/list/card_component.js [hunk 9]
```diff
@@ -27,7 +27,7 @@ export default class ReplicaSetCardController {
    * @param {!./../../common/namespace/service.NamespaceService} kdNamespaceService
    * @ngInject
    */
-  constructor($state, $interpolate, kdNamespaceService) {
+  constructor($state, $interpolate, kdNamespaceService, kdScaleService) {
```
Comment (floreks) [inline src/app/frontend/replicaset/list/card_component.js]: missing annotation

File: src/app/frontend/job/list/card_component.js [hunk 10]
```diff
@@ -26,7 +26,7 @@ export default class JobCardController {
    * @param {!./../../common/namespace/service.NamespaceService} kdNamespaceService
    * @ngInject
    */
-  constructor($state, kdNamespaceService) {
+  constructor($state, kdNamespaceService, kdScaleService) {
```
Comment (floreks) [inline src/app/frontend/job/list/card_component.js]: missing annotation

File: src/app/frontend/job/detail/actionbar_controller.js [hunk 11]
```diff
@@ -20,8 +20,17 @@ export class ActionBarController {
    * @param {!backendApi.JobDetail} jobDetail
    * @ngInject
    */
-  constructor(jobDetail) {
+  constructor(jobDetail, kdScaleService) {
```
Comment (floreks) [inline src/app/frontend/job/detail/actionbar_controller.js]: missing annotation

File: src/app/frontend/deployment/list/card_component.js [hunk 12]
```diff
@@ -42,6 +42,11 @@ export default class DeploymentCardController {
 
     /** @private {!./../../common/namespace/service.NamespaceService} */
     this.kdNamespaceService_ = kdNamespaceService;
+
+    /** @private
+     * {!./../replicationcontrollerdetail/replicationcontroller_service.ReplicationControllerService}
```
Comment (floreks) [inline src/app/frontend/deployment/list/card_component.js]: wrong annotation

File: src/app/frontend/deployment/list/card_component.js [hunk 13]
```diff
@@ -27,7 +27,7 @@ export default class DeploymentCardController {
    * @param {!./../../common/namespace/service.NamespaceService} kdNamespaceService
    * @ngInject
    */
-  constructor($state, $interpolate, kdNamespaceService) {
+  constructor($state, $interpolate, kdNamespaceService, kdScaleService) {
```
Comment (floreks) [inline src/app/frontend/deployment/list/card_component.js]: missing annotation

**Comment by maciaszczykm** [comment]
Please add new commits instead of rebasing, so it will be easier for us to check new changes.

**Comment by kenan435** [comment]
@floreks, @maciaszczykm  could one of you guys have a look at the pr again. I am pretty sure all of the requested changes have been addressed.

**Comment by cheld** [comment]
@kenan435 write PTAL when you are done and PR is ready for review

**Pull request merged**